cloud/google/gce_mig.py

```diff
@@ -0,0 +1,763 @@
+    _ = Provider.GCE
+    HAS_LIBCLOUD = True
+except ImportError:
+    HAS_LIBCLOUD = False
+
+try:
+    from ast import literal_eval
+    HAS_PYTHON26 = True
+except ImportError:
+    HAS_PYTHON26 = False
+
+
+def _check_params(params, field_list):
+    """ 
+    Helper to validate params.
+
+    Use this in function definitions if they require specific fields
+    to be present.
+
+    :param params: structure that contains the fields
+    :type params: ``dict``
+
+    :param field_list: list of dict representing the fields
+                       [{'name': str, 'required': True/False', 'type': cls}]
+    :type field_list: ``list`` of ``dict``
+
+    :return True, exits otherwise
+    :rtype: ``bool``
+    """
+    for d in field_list:
+        if not d['name'] in params:
+            if d['required'] is True:
+                return (False, "%s is required and must be of type: %s" %
+                        (d['name'], str(d['type'])))
+        else:
+            if not isinstance(params[d['name']], d['type']):
+                return (False,
+                        "%s must be of type: %s" % (d['name'], str(d['type'])))
+
+    return (True, '')
+
+
+def _validate_autoscaling_params(params):
+    """ 
+    Validate that the minimum configuration is present for autoscaling.
+
+    :param params: Ansible dictionary containing autoscaling configuration
+                   It is expected that autoscaling config will be found at the
+                   key 'autoscaling'.
+    :type  params: ``dict``
+
+    :return: Tuple containing a boolean and a string.  True if autoscaler 
+             is valid, False otherwise, plus str for message.
+    :rtype: ``(``bool``, ``str``)``
+    """
+    if not params['autoscaling']:
+        # It's optional, so if not set at all, it's valid.
+        return (True, '')
+    if not isinstance(params['autoscaling'], dict):
+        return (False,
+                'autoscaling: configuration expected to be a dictionary.')
+
+    # check first-level required fields
+    as_req_fields = [
+        {'name': 'name', 'required': True, 'type': str},
+        {'name': 'enabled', 'required': True, 'type': bool},
+        {'name': 'policy', 'required': True, 'type': dict}
+    ] # yapf: disable
+
+    (as_req_valid, as_req_msg) = _check_params(params['autoscaling'],
+                                               as_req_fields)
+    if not as_req_valid:
+        return (False, as_req_msg)
+
+    # check policy configuration
+    as_policy_fields = [
+        {'name': 'max_instances', 'required': True, 'type': int},
+        {'name': 'min_instances', 'required': False, 'type': int},
+        {'name': 'cool_down_period', 'required': False, 'type': int}
+    ] # yapf: disable
+
+    (as_policy_valid, as_policy_msg) = _check_params(
+        params['autoscaling']['policy'], as_policy_fields)
+    if not as_policy_valid:
+        return (False, as_policy_msg)
+
+    # TODO(supertom): check utilization fields
```

Comment:
Yes.  The utilization will be be passed through as-is and the API will accept/reject.  The TODO is for us to one day, save the API call and validate in the module.